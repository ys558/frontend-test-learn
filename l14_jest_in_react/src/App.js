import React from 'react';

function App() {
  return (
    <div className="app" title="123" data-test="app-container">
      1234
      <nav className="header">header</nav>
      <div className="content">
        <div className="sidebar-left">sidebar left</div>
        content
      </div>
      <footer className="footer">footer</footer>
    </div>
  );
}

export default App;
