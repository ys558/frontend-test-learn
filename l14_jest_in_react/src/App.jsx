import React from 'react';
import TodoList from './containers/TodoList';
import Counter from './containers/Counter/MyCounter';
import Pokemon from './containers/Pokemon/Pokemon';
import TestHookByEnzyme from './containers/TestHookByEnzyme/TestHookByEnzyme'

function App() {
  return (
    <div>
      <h1>Demo 1</h1>
      <Counter/>
      <h1>Demo 2</h1>
      <TodoList/>
      <h1>Demo 3 test hooks by @testing-library/react-hooks</h1>
      <Pokemon />
      <h1>Demo 4 test hooks by Enzyme</h1>
      {/* <TestHookByEnzyme/> */}
    </div>
  );
}

export default App;
