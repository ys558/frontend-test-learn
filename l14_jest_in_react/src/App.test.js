import React from 'react';
import App from './App';

// 这里引入Enyme，他对React组件进行包装：
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

test('renders without crashing', () => {
  // 由于create-react-app本身是单页面应用，如果要对其进行class里的状态进行检查，普通ReactDOM操作不合适：
  // const div = document.createElement('div')
  // ReactDOM.render(<App/>, div)
  // const container = div.getElementsByClassName('App')
  // console.log(container)
  // expect(container.length).toBe(1)

  // shallow是浅渲染: 适用于单元测试: 即组件无嵌套时使用:
  const wrapperShallow = shallow(<App/>)
  // console.log(wrapperShallow.find('.app'))
  expect(wrapperShallow.find('.app').length).toBe(1)
  // console.log(expect(wrapperShallow.find('.app').prop('title')))
  expect(wrapperShallow.find('.app').prop('title')).toBe('123')

  /*
  另外, 如果我们用源代码的class去作为选择器进行选择, 耦合性会比较高,
  假如源代码改变class名字,那么所有测试代码也要一并跟着改.
  解决办法:
  在源代码设置一个转为前端测试的属性名, 如: data-test
  */
  expect(wrapperShallow.find('[data-test="app-container"]').length).toBe(1)
  expect(wrapperShallow.find('[data-test="app-container"]').prop('title')).toBe('123')

  // jest-enzyme 的匹配器:
  // npm i -D jest-enzyme 或 yarn add -D jest-enzyme
  // https://github.com/FormidableLabs/enzyme-matchers/tree/master/packages/jest-enzyme#usage-with-create-react-app
  // 拉到最底部有配置介绍, 在create-react-app中的 src/setupTests.js 文件里添加:
  // import 'jest-enzyme';
  // 之后能能运用其提供的api, 把上面的两行改写:
  expect(wrapperShallow.find('[data-test="app-container"]')).toExist()
  expect(wrapperShallow.find('[data-test="app-container"]')).toHaveProp('title','123')

  // mount测试一堆组件包括子组件的集合, 适用于集成测试: 即组件存在嵌套时使用:
  const wrapperMount = mount(<App/>)
  // 如果页面涉及敏感关键信息, 不宜过度更改的话, 可以用snapshot的功能, 把快照保存下来: 
  // 生产 快照, 如果改了源码, 就会出现测试不通过, 这时如果要更新快照, 则在控制台按u键
  // l14_jest_in_react\src\__snapshots__\App.test.js.snap
  expect(wrapperMount).toMatchSnapshot()
  
});
