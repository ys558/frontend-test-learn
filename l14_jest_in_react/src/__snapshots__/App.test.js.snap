// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<App>
  <div
    className="app"
    data-test="app-container"
    title="123"
  >
    1234
    <nav
      className="header"
    >
      header
    </nav>
    <div
      className="content"
    >
      <div
        className="sidebar-left"
      >
        sidebar left
      </div>
      content
    </div>
    <footer
      className="footer"
    >
      footer
    </footer>
  </div>
</App>
`;

exports[`renders without crashing 2`] = `
<div
  className="app"
  data-test="app-container"
  title="123"
>
  1234
  <nav
    className="header"
  >
    header
  </nav>
  <div
    className="content"
  >
    <div
      className="sidebar-left"
    >
      sidebar left
    </div>
    content
  </div>
  <footer
    className="footer"
  >
    footer
  </footer>
</div>
`;
